src/Stream.ts

```diff
@@ -40,30 +40,26 @@ export const switchMap = <E, X, E2>(fx: Fx.Fx<E, X>, f: (a: Event<E>) => Fx.Fx<E
     })
   )
 
-export const withEmitter = <A>(f: (emitter: Emitter<A>) => Disposable): Fx.Fx<Async.Async | Stream<A>, void> =>
-  Fx.fx(function* () {
-    const queue = new Queue.UnboundedQueue<A>()
-
-    const disposable = f({
-      event(a) { queue.offer(a) },
-      end() { dispose(queue) }
-    })
+export const fromQueue = <A>(queue: Queue.Take<A>): Fx.Fx<Async.Async | Stream<A>, void> => Fx.fx(function* () {
+  const take = Queue.take(queue)
 
-    const take = Async.run(() => queue.take())
+  while (!queue.disposed) {
+    const next = yield* take
+    if (next.tag === 'fx/Queue/Taken') yield* event(next.value)
+  }
+})
 
-    while (!queue.disposed) {
-      const next = yield* take
-      if (next.tag === 'fx/Queue/Take') yield* event(next.value)
-    }
+export const withEmitter = <A>(
+  f: (o: Queue.Offer<A>) => Disposable,
+  q: Queue.Queue<A> = new Queue.UnboundedQueue<A>()
+): Fx.Fx<Async.Async | Stream<A>, void> =>
```

Comment:
This now uses the Queue.Offer interface directly rather than introducing an Emitter middle-man object.

Reply:
Maybe a new name is order then? It's no longer really "with an emitter".

Maybe `withOffer` or `withQueue` or something similar. 

Maybe `withEnqueue` if we also adjusting the naming of Offer to Enqueue

Reply:
Good thoughts!  I'm fine with any pair of names (for the methods and the respective interfaces) that seems intuitive, e.g. offer/take, put/take, enqueue/dequeue, push/shift.  Any thoughts on those?  Others?

I like the idea of renaming withEmitter to align with whatever names we pick 👍

Reply:
Enqueue/dequeue seem like the most obvious names.  I'll update this PR to use them and we can see how we like it.

Another thought on withEmitter is that it could potentially go away.  if you create a Queue, you can hand it to both the producer function and Stream.fromQueue yourself.  Or, we could still keep it (or add it back later) as a convenience.

Reply:
👍 I like enqueue/dequeue too

I usually find this operator to be one of the core features to adapt external world APIs into the world of Stream. I use it relatively seldomly directly, but it does feel essential when you have APIs like `EventEmitter.on` or `EventTarget.addEventListener`

Reply:
I was a little surprised by this, but `withEnqueue` (renamed from withEmitter) is now just a trivial bracket around `fromDequeue`, so I don't feel strongly about keeping it or removing it.  We can keep it for now.